### PR TITLE
feat(lint): add options from `typescript-eslint` to `useUnifiedTypeSignature`

### LINKED ISSUE #8355: 📎 Add options to `useUnifiedTypeSignature`
### Description

We should port over the [`ignoreDifferentlyNamedParameters`](https://typescript-eslint.io/rules/unified-signatures/#ignoredifferentlynamedparameters) and [`ignoreOverloadsWithDifferentJSDoc`](https://typescript-eslint.io/rules/unified-signatures/#ignoreoverloadswithdifferentjsdoc) options from the source rule. I don't really have any strong feelings about the naming for these options, even though they are a bit verbose. Makes some sense to keep the same option names.

Discussed in #7218

Since these are new options, they have to go in a minor release. Please make sure to make your PR against the `next` branch.

**Want to contribute?** Let us know you are interested! We will assign you to the issue to prevent several people to work on the same issue. Don't worry, we can unassign you later if you are no longer interested in the issue! Read our [contributing guide](https://github.com/biomejs/biome/blob/main/CONTRIBUTING.md) and [analyzer contributing guide](https://github.com/biomejs/biome/blob/main/crates/biome_analyze/CONTRIBUTING.md).

### PR DESCRIPTION
**Pull request by Bertie690**
## Summary

Fixes #8355 

Adds 2 new options from TSEslint to `useUnifiedTypeSignature` that cause the rule to ignore function overloads with either differently named parameters or different JSDoc comments.

The basic gist of the code is as follows:
1. Moved the giant nested for loop into its own function, `try_merge_overloads`, to reduce nesting and add more encapsulation.
2. Added conditionals to said function to check for identical JSDocs and/or parameter names, respectively. The former is done by exposing an underlying iterator from the existing JSDoc parsing service code, while the latter is done with a hand-rolled `same_param_names` function for `AnyJsParameterListExt` that compares parameter names piecewise.

> [!IMPORTANT]
> I cannot stress how painful it was to attempt to get a parameter's name from an AST node. The amount of duplicated code and other name equality things I came across makes me really worried at what other sorts of duplication could be lurking in the darkness. Cleaning up the various `NameEquals` vs `IsNameEqual` instances is above my paygrade, but would make for a good next refactor or 2.
> (As an aside, I think the various syntax token unions could use a bit of abstraction over their internal details. I arguably shouldn't need to go digging into the generated code inside `biome_js_syntax` for what types of nodes make up an `AnyParameter` or `AnyJsBinding` just to get the name of a damn thing.)

## Test Plan
Created several new test files for the various new options, including both valid and invalid examples.

I also took it upon me to add even more stupidly obscure spec tests for things like array spread and object literals that we did not test prior. (I like edge cases)

## Docs

Added rust docs for the new options.

### CODE REVIEW

File: .changeset/rare-maps-read.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
+---
+"@biomejs/biome": minor
+---
+
+Added 2 options from `typescript-eslint` (`ignoreDifferentlyNamedParameters` and `ignoreDifferentJsDoc`) to `lint/style/useUnifiedTypeSignatures`.
```
Comment (dyc3) [inline .changeset/rare-maps-read.md]: nit:
```suggestion
Added 2 options from `typescript-eslint` (`ignoreDifferentlyNamedParameters` and `ignoreDifferentJsDoc`) to [`useUnifiedTypeSignatures`](https://biomejs.dev/linter/rules/use-unified-type-signatures/).
```

File: crates/biome_js_analyze/src/lint/style/use_unified_type_signatures.rs [hunk 2]
```diff
@@ -73,6 +91,58 @@ declare_lint_rule! {
...
+    ///
+    /// ```ts,use_options
+    /// function bake(numApples: number): void;
+    /// function bake(cakeType: string): void;
+    /// ```
+    ///
+    /// ### `ignoreDifferentJsDoc`
+    ///
+    /// If set to `true`, overloads with different JSDoc comments from one another will be ignored.
+    /// Ones without comments will still be checked (and can poentially be merged with documented signatures)
```
[the remark this answers is not in the document]
Comment (dyc3) [reply crates/biome_js_analyze/src/lint/style/use_unified_type_signatures.rs]: this is valid

File: crates/biome_rule_options/src/use_unified_type_signatures.rs [hunk 3]
```diff
@@ -1,6 +1,18 @@
...
-pub struct UseUnifiedTypeSignaturesOptions {}
+pub struct UseUnifiedTypeSignaturesOptions {
+    /// Whether to ignore overloads with differently named parameters.
+    #[serde(skip_serializing_if = "Option::<_>::is_none")]
+    pub ignore_differently_named_parameters: Option<bool>,
+
+    /// Whether to ignore overloads with different JSDoc comments.
+    #[serde(skip_serializing_if = "Option::<_>::is_none")]
+    pub ignore_different_jsdoc: Option<bool>,
+}
```
Comment (dyc3) [inline crates/biome_rule_options/src/use_unified_type_signatures.rs:17]: nit: Consider adding an impl block for these so you don't have to write `unwrap_or_default()` when using them.

```rust
impl UseUnifiedTypeSignaturesOptions {
  fn ignore_differently_named_parameters() -> bool { ... }
  fn ignore_different_jsdoc() -> bool { ... }
}
```
Plus, it makes it way less error prone to change the defaults if we decide to do that later.

File: crates/biome_js_analyze/tests/specs/style/useUnifiedTypeSignatures/invalidJsdoc.options.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{
+    "$schema": "../../../../../../packages/@biomejs/biome/configuration_schema.json",
+    "linter": {
+        "rules": {
+            "style": {
+                "useUnifiedTypeSignatures": {
+                    "level": "on",
+                    "options": {
+                        "ignoreDifferentJsDoc": true
```
[the remark this answers is not in the document]
Comment (Bertie690) [reply crates/biome_js_analyze/tests/specs/style/useUnifiedTypeSignatures/invalidJsdoc.options.json:9]: this is consistent with existing rules and is shorter

File: .changeset/rare-maps-read.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * Generate a random battle with the specified starters.
+ */
+export async function startBattle(starters: SpeciesId[]): Promise<void>;
+/**
+ * Generate a random battle with 3 randomly picked starters.
+ * @deprecated - specify the starters to avoid RNG and be more explicit
+ */
+// TODO: Remove
+export async function startBattle(): Promise<void>;
```
Comment (dyc3) [inline .changeset/rare-maps-read.md]: I would like to see these code examples be less verbose, especially because they'll be in the changelog.

**Review by dyc3 — COMMENTED** [review]
Notes mostly on the changeset. Everything else looks good!

**Comment by Bertie690** [comment]
About the CI failure - I did not touch that file at all, so I'm assuming the test is borked or something.
IDK why a random vue test would be crashing and it's certainly not my fault lol

**Review by dyc3 — APPROVED** [review]
